File: unity/Assets/Scripts/unityagent/SystemMessage/EgressSystemMessage.cs
using Unity.VisualScripting;
using UnityAgent.Util.Serialization;

namespace UnityAgent.SystemMessage
{
public class EgressSystemMessage
{
    private readonly RoutingInfo routingInfo;
    private readonly SystemMessageMetadata metadata = null;
    private readonly byte[] data = new byte[0];

    public EgressSystemMessage(RoutingInfo routingInfo)
    {
        this.routingInfo = routingInfo;
    }

    public EgressSystemMessage(RoutingInfo routingInfo,
                               SystemMessageMetadata metadata)
    {
        this.routingInfo = routingInfo;
        this.metadata = metadata;
    }

    public EgressSystemMessage(RoutingInfo routingInfo,
                               SystemMessageMetadata metadata,
                               byte[] data)
    {
        this.routingInfo = routingInfo;
        this.metadata = metadata;
        this.data = data;
    }

    public SystemMessageMetadata Metadata => metadata;
    public RoutingInfo RoutingInfo => routingInfo;

    public byte[] Serialize()
    {
        byte[] sMetadata = metadata == null ? new byte[0] : metadata.Serialize();

        int sMetadataSize = sMetadata.Length;
        byte[] metadataSizeFlag = NetworkDataSerializer.IntToBytes(sMetadataSize);
        int metadataSizeFlagSize = metadataSizeFlag.Length;

        int sDataSize = data.Length;
        byte[] dataSizeFlag = NetworkDataSerializer.IntToBytes(sDataSize);
        int dataSizeFlagSize = dataSizeFlag.Length;

        int totalMessageSize = metadataSizeFlagSize + sMetadataSize +
                               dataSizeFlagSize + sDataSize;
        byte[] sRoutingInfo = SerializeRoutingInfo(totalMessageSize);
        int sRoutingInfoSize = sRoutingInfo.Length;

        int systemMessageSize = sRoutingInfoSize + totalMessageSize;
        byte[] systemMessage = new byte[systemMessageSize];
        
        int offset = 0;
        System.Buffer.BlockCopy(sRoutingInfo, 0, systemMessage, 0,
                                sRoutingInfoSize);
        offset += sRoutingInfoSize;
        System.Buffer.BlockCopy(metadataSizeFlag, 0, systemMessage, offset,
                                metadataSizeFlagSize);
        offset +=  metadataSizeFlagSize;
        System.Buffer.BlockCopy(sMetadata, 0, systemMessage, offset,
                                sMetadataSize);
        offset += sMetadataSize;
        System.Buffer.BlockCopy(dataSizeFlag, 0, systemMessage, offset,
                                dataSizeFlagSize);
        offset += dataSizeFlagSize;
        System.Buffer.BlockCopy(data, 0, systemMessage, offset, sDataSize);

        return systemMessage;
    }

    private byte[] SerializeRoutingInfo(int totalMessageSize)
    {
        routingInfo.TotalMessageSize = totalMessageSize;
        byte[] sRoutingInfo = routingInfo.Serialize();
        return sRoutingInfo;
    }
}
}

File: ros/ros_packages/rosagent/rosagent/message/egress_system_message.py
from rosutility.message.serializer import serializer as Serializer

class EgressSystemMessage:

    def __init__(self, routing_info, metadata=None, data:bytes=None):
        self._routing_info = routing_info
        self._metadata = metadata
        self._data = bytes() if data is None else data

    def serialize(self):
        s_metadata = (bytes() if self._metadata is None
                      else self._metadata.serialize())

        metadata_length = len(s_metadata)
        s_metadata_length = Serializer.uint_to_bytes(metadata_length)

        data_length = len(self._data)
        s_data_length = Serializer.uint_to_bytes(data_length)
        
        total_message_size = (len(s_metadata_length) + metadata_length +
                              len(s_data_length) + data_length)
        s_routing_info = self._serialize_routing_info(total_message_size)

        return (s_routing_info + s_metadata_length + s_metadata + 
                s_data_length + self._data)

    def _serialize_routing_info(self, total_message_size):
        self._routing_info.set_total_message_size(total_message_size)
        return self._routing_info.serialize()
